### DISCUSSION TITLE
Install joplin-vieweb in parallel of other web apps using apache2?

### DISCUSSION THREAD

**Discussion by Eliness** [joplin-vieweb/joplin-vieweb#45] [Q&A] [answered]
Hello Nicolas,

First of all, thank you so much for developing and maintaining joplin-vieweb - it is the only thing that I am currently missing from Joplin, and being able to access my notes through a web browser would be a real improvement in my everyday life!

I currently have a small server at home, running different web apps using apache2. For example, it hosts a Nextcloud instance (which coincidently also serves as a sync point for my Joplin notes).

I understand that joplin vieweb uses nginx, and that the docker sets it up all nicely. I have several questions here :
- Can I use the docker container that would install nginx in parallel of apache2?
- How do I deal with the fact that the ports 80 and 443 are already used by apache2?
- Can I install joplin-vieweb in a subfolder, eg https://www.mydomain.com/joplinvieweb?

I'm not very skilled in syadmin nor webhosting, and using docker is new to me so apologies in advance if I'm mixing up some concepts. Willing to learn :)

Thank you for your time!

**Comment by gri38** [comment]
Hello @Eliness .
Thanks for your feed backs. 

I'm not very skill neither with webhosting ;-) but we'll manage to setup your server ;-)

First thing: yes you'll be able to install joplin vieweb with your current web server (I hope you won't have to install an upstream reverse proxy and apache can handle that).

The docker containers involved in JVW are:
![image](https://user-images.githubusercontent.com/26554495/211004383-d951f67d-100d-44bf-90c1-7bc4f5a31d1f.png)

### Container 1️⃣
It is optional. It is the "front reverse proxy": request from your webbrowser client will arrive on it.
Its purpose is:
- To manage the certificates (for https).
- "Translate" https to http, and send the request to container 2️⃣
- Same for the way back (from joplin app to your webbrowser client)
### Container 2️⃣
It is the webserver for JVW application. (Webserver = daphne).
I could have left nginx out of this docker image, but I choose nging to serve some static content. I could have done differently (we don't really care, it's just to explain the nginx purpose in this container).

And briefly
### Container 3️⃣
It is optional too: its purpose is to get and renew certificates for container 1️⃣ nging. But if you already handle certificates through apache2, you don't need 1️⃣nor 3️⃣.
### Container 4️⃣
It runs the joplin app (joplin terminal).
And I extended joplin terminal app with a few features need by application in container 2️⃣.
In this container, the only purpose of nginx it to make calls to joplin terminal with localhost as origin (it's required by joplin terminal API)

# SO
For your configuration, you should end up to sthg like that:
![image](https://user-images.githubusercontent.com/26554495/211008194-04281249-11d0-4187-9405-8e39ba58c7a3.png)
Correct me if I'm wrong (I don't know about apache)
- You apache server receive the https requests, handle the certificates
- Your apache 2 conf probably configure that everything like https://yourdomaine**/nextcloud** is handle by php and the php site path is /var/www/nextcloud (or something like that)
- So you have to find how to configure apache to say that everything like https://yourdomain**/joplin** should be send to port 8000 of host django-joplin-vieweb. Maybe this [web page](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/How-to-configure-Apache-as-a-reverse-proxy-example) can help (disclamer: I am not the man on this page and I don't know him ;-))

As you can see, no 80 or 443 port conflict: since you already have a server, you don't need container 1 and 3.
So the docker-compose file to use as a starting point is [docker-compose-joplin-only.yml](https://github.com/joplin-vieweb/joplin-vieweb/blob/main/docker-compose-joplin-only.yml). Read the doc in Installation paragraph [here](https://github.com/joplin-vieweb/joplin-vieweb/#installation).

Then, either you'll have to start the containers manually (or script). Or maybe you can configure apache to do so, I don't know if it's possible.
The "best" would be your apache server is in a docker image, and you could set it in the docker-compose file, so everything is started together.
If it's not the case today, I won't advise to change your current setup. 
=> Do you use apache in a docker container ?

that's it for the moment, we need to iterate o your configuration.

**Comment by Eliness** [comment]
Awesome detailed explanation! The drawings really helped me understand the container logic, thank you so much :heart: 

You understood things right when it comes to how apache works! I don't use it in a docker container, but I don't mind starting things manually when required. However, before trying proper domain redirection, I think I need to make things work locally first :) 

I understood I don't need container 1 nor 3, since these things are already managed by my server. So I'm focusing on containers 2 and 4, using `docker-compose-joplin-only.yml.`

I completed the `ORIGINS` field with the local IP address of the server + the 8080 port, as following :
```
x-common-variables: &common-variables
   ORIGINS: "'http://localhost:8080', 'http://192.168.1.163:8080'"
```

I don't really know what to put in the `ports` field, I have no proxy - I tried deleting the parameter entirely, or just setting it up as 8000, I have honestly no idea what I'm doing. Still, in both cases I could get docker to set up and run nicely so it's a first step. See the last lines of the console here: 
```
...
django-joplin-vieweb_1  | ALLOWED_HOSTS =  ['localhost', '192.168.1.163']
django-joplin-vieweb_1  | CSRF_TRUSTED_ORIGINS =  ['http://localhost:8080', 'http://192.168.1.163:8080']
django-joplin-vieweb_1  | Starting server at tcp:port=8001:interface=127.0.0.1
django-joplin-vieweb_1  | HTTP/2 support not enabled (install the http2 and tls Twisted extras)
django-joplin-vieweb_1  | Configuring endpoint tcp:port=8001:interface=127.0.0.1
django-joplin-vieweb_1  | Listening on TCP address 127.0.0.1:8001
joplin-terminal-xapi_1  | Synch finished: return code: 0
joplin-terminal-xapi_1  |     status: 2023-01-06T18:10:04.083Z
joplin-terminal-xapi_1  |     stdout: Not authentified with (None). Please provide any missing credentials.
joplin-terminal-xapi_1  | 
joplin-terminal-xapi_1  |     stderr: 
```

So far, I get a `ERR_CONNECTION_REFUSED` when trying to access `http://192.168.1.163:8080` once the container is up. Connecting to the server, I can access my other pages using `localhost`, but not joplin-vieweb using `localhost:8080`. This is where I am stuck for now, is it a question of configuration (proxy field ?) 

Hope you don't mind explaining things to me like I'm 5, I feel very lost at the moment :laughing: If it helps, I'd be happy to contribute to the documentation in return once I think I got things right :)

**Comment by gri38** [comment]
Hi.

## About your origins
Just to check: it means your apache server is listening to port 8080, right ? If so, you did the correct thing for ORIGINS

## About ports in the docker-compose file
As you can see in the above schema, the nginx in container django-joplin-vieweb must be called on port 8000.
By default, containers doesn't expose ports to host machine (your server). So in order your apache, _running on the host machine_, to be able to contact django-joplin-vieweb, the later must expose a port.
It is the purpose of ports directive in the docker-compose file, under the service django-joplin-vieweb.
On your host machine:
-  1️⃣ If port 8000 is available, you can expose it:
```yaml
django-joplin-vieweb:
    image: gri38/django-joplin-vieweb:latest
    ports:
      - 8000:8000
```
And then, on your server, if you input `http://localhost:8000`, it will reach the container on port 8000 (and so will reach nginx).
-  2️⃣  If port 8000 is **not** available (it's quite common that port 8000 is used on linux system), you can expose another port not used by your server yet, let's say 1234:
```yaml
django-joplin-vieweb:
    image: gri38/django-joplin-vieweb:latest
    ports:
      - 1234:8000
```
And then, on your server, if you input `http://localhost:1234`, it will reach the container on port 8000 (and so will reach nginx).
![image](https://user-images.githubusercontent.com/26554495/211102779-fee0c7cb-0f6c-45e8-a5a6-e18107220146.png)

The "routing" will be:
- on your browser, you input http://localhost:8080/joplin
- it will reach your apache, which you have configured to forward to: [http://localhost:1234[or 8000]/joplin](#)
- and that will reach the nginx container: all's right.

## About the remaining errors
- Once everything is setup (on the server), don't forget to follow the [Configuration and usage §](https://github.com/joplin-vieweb/joplin-vieweb/#configuration-and-usage).
  - First, you'll have to create at least one user: [http://localhost:8080/**admin**](http://localhost:8080/admin) 
  - And when it comes to login into joplin viweb application, don't forget to add joplin in the url, as explaine in the doc: [http://localhost:8080/**joplin**](http://localhost:8080/joplin).
  - Then you can configure your sync settings... and try the app.
- After all that if you still have some issues, you could try to add 'http://0.0.0.0:8080', 'http://127.0.0.1:8080''' in your origins (even if I don't think it's needed)

## Once again: /joplin at the end (or /admin to configure users)
http://localhost:8080 is not suppose to do anything.
The urls of my app are: [http://localhost:8080/**admin**](http://localhost:8080/admin) and [http://localhost:8080/**joplin**](http://localhost:8080/joplin)

# Be careful
There is no reason, but backup your notes before all that. 
My app is safe about that: I don't do anything with the synchro: I totally rely on joplin terminal.
But in the forum, we quite often read about users of joplin that made some mistakes and erase all their notes: backup during this kind of setup.

Be brave, and let me know about your results.

**Comment by Eliness** [comment]
Hey Nicolas,

Your detailed answers are incredible, thank you so much for the time you spent! It really helps me understand the logic and how things connect to eachother!

Making slow progress here :blush: I still have no clue what I'm doing with ports, I don't know how to redirect the origin one (1234 in your example) to the container one (8000) using Apache so I chose the lazy option for now: setting both of them to 1234, and it seems to work. Indeed I can connect from the server to `localhost:1234/admin` and from another local browser to  `serverip:1234/admin`. So I can configure everything following your instructions and everything works locally: I could create a user and then, going to `serverip:1234/joplin`, set things up properly: my notes are currently syncing :tada: 

Now for remote access, I understand I need to configure Apache and use a reverse proxy to go from the external `https://domain.net/joplin` to my internal `http://serverip:1234/joplin` . I've been trying unsuccessfully for the past hour, mainly because of issues with SSL ; I understand however that you are not versed in Apache, so I'll scout some documentation on my own. I'll come back with the answer if/when I find it!

**Comment by gri38** [comment]
Hi.

Cool, you're almost done.

About ports, read carefully my explanation: the "redirection" is done in the ports line in docker-compose: if you input 1234:8000, it means that anything on your server that targets localhost:1234 will be "catch" by docker and sent into the container on port 8000.

About your apache conf, you said "...and use a reverse proxy to go from the external https://domain.net/joplin to my internal http://serverip:1234/joplin": yes. But I think you go from what you said to `http://localhost:1234/joplin`, since your apache and docker are on the same server (are they?)

**Comment by Eliness** [comment, marked as answer]
After quite some struggle with Apache and domain redirections this weekend, everything seems to be working smoothly the way I want it to 🎉🎉🎉 The link you shared about reverse proxies in Apache greatly helped as a starting point, thank you again for that! Long story short I could not figure out how to properly redirect things using subfolders (`https://domain.net/joplin`) , so I set up a subdomain instead (`https://sub.domain.net/`) and an associated VirtualHost file - this way I can access both `https://sub.domain.net/joplin` and `https://sub.domain.net/admin`.

I think I understood the nuance about the ports thanks to your extra explanation - I didn't change my original setup which was working as is!
And you were right, of course the subdomain redirection needed to be to localhost since everything is on the same machine 🙃 

One remaining question: is there a way to change the parameters originally set in the yaml file (eg ALLOWED_HOSTS) without having to delete the volumes / reinstall everything? (I'm very ignorant about docker 😅) I tried to look into the volumes to see if I could find files where to edit the parameters directly, but could not find a file where they were used.

Finally,  I wrote down below a summary of all the steps I ran to make things work!

Thanks again for all your help ❤️ 

# Install joplin-vieweb on a local web server, and access it externally from a subdomain through apache2

❗❗❗ **MEGA DISCLAIMER**: I'm quite a beginner in all of this so surely some of these instructions are approximate if not wrong. I'm far from claiming the following process is secure nor optimal. This is only how I got things to work, any suggestion for improvement is more than welcome :) ❗❗❗ 

The context is: you have a selfhosted web server with apache2 running, and you want to install joplin-vieweb on it and access it from an external subdomain (https://sub.domain.net in the following instructions).

# Install joplin-vieweb on your server through docker
Download the [docker-only yaml file](https://github.com/joplin-vieweb/joplin-vieweb/blob/main/docker-compose-joplin-only.yml) and edit the following parameters:
- Line 4: ORIGINS: Adresses from which you want to be able to connect to joplin-vieweb.
`ORIGINS: "'http://localhost', 'http://localipofserver','https://sub.domain.net"`
- Line 15: ports: two port numbers. The first one of your choice (I used 1234), is the one on your server which will be redirected to the second one in your container.
- ` - 1234:8000`

Then on the command line in the same folder than the yaml file, run the containers in detached mode:
`docker-compose up -d`

On your web server, check that you can access `http://localhost:1234/admin`
From your local network, check that you can access `http://localserverip:1234/admin`

# Set joplin-vieweb up
Follow the instructions of [Configurations and usage](https://github.com/joplin-vieweb/joplin-vieweb/#configuration-and-usage) in the doc.

Once the sync is done, if your notes are encrypted they should not display properly. You need to run the decryption using the following command line from your server:
`docker exec -ti joplin-vieweb_joplin-terminal-xapi_1 joplin e2ee decrypt`
It will ask for your Joplin master key and will take some time to decrypt the files. It happened to me that the decryption interrupted and returned an error because Joplin started syncing at the same time. If so, just wait for the sync to finish then re-run the decryption command.

You should now be able to access your notes on your local network using `http://localserverip:1234/joplin`

# Access your notes from the outside: Create an apache2 VirtualHost file which will redirect your subdomain
Let's assume your subdomain is: http://sub.domain.net (we'll tackle https later).

In `/etc/apache2/sites-available`, create the following `subdomain.conf` file:

```
<VirtualHost _default_:80>
ServerName sub.domain.net
ProxyPreserveHost On
<Location />
        ProxyPass http://localhost:1234/
        ProxyPassReverse http://localhost:1234/
        Require all granted
        Options none
</Location>
</VirtualHost>
```

Then relaunch apache2 with the following: `sudo systemctl restart apache2`

# Redirect your subdomain to the server IP
This depends on your domain provider service, there should be a parameter in the administration where you can use an A redirection to redirect `http://sub.domain.net` to the **public** ip address of your server. To get the latter, I used the following command line: `curl icanhazip.com`.

Now you should be able to access joplin-vieweb from anywhere on the following address:  `http://sub.domain.net/joplin`

# (optional) Generate an SSL certificate and edit the VirtualHost so that joplin-vieweb is accessible through https
You can use certbot to generate an SSL certificate as following: `sudo certbot certonly -d sub.domain.net` . The path to your keys should be returned as output.
Then you need to edit the  `subdomain.conf` file accordingly (note changes in the port of the first line, the added lines and the paths to your keys)

```
<VirtualHost _default_:443>
ServerName sub.domain.net
ProxyPreserveHost On
<Location />
        ProxyPass http://localhost:1234/
        ProxyPassReverse http://localhost:1234/
        RequestHeader set X-Forwarded-Protocol "https"
        Require all granted
        Options none
</Location>
SSLProxyEngine on
SSLCertificateFile /etc/letsencrypt/live/sub.domain.net/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/sub.domain.net/privkey.pem
</VirtualHost>
```

You should now be able to access joplin-vieweb from anywhere through `https://sub.domain.net/joplin`

**Comment by gri38** [comment]
Hi.
I'm glad you succeeded !

Thanks for your doc, I've integrated it in the install documentation! [here](https://github.com/joplin-vieweb/.github/blob/main/profile/README.md#features--install--releases) and [here](https://github.com/joplin-vieweb/joplin-vieweb/blob/main/README.md#installation)
And about ALLOWED_HOSTS changes not processed: I have a but on it: https://github.com/joplin-vieweb/joplin-vieweb/issues/34
Now your the second reporting that, and indeed when you are trying to make everything works, it's annoying. I'll fix it for the next person ;-)
